oap-ape/ape-java/ape-hcfs/src/main/java/com/intel/oap/fs/hadoop/ape/hcfs/CacheAwareFileSystem.java

```
@@ -0,0 +1,297 @@
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * A hadoop file system implementation that wraps HDFS data accessing
+ * and provides locations of cached data.
+ *
+ * Locations of cached data can be loaded from redis.
+ *
+ * Usage: set fs.hdfs.impl to com.intel.oap.fs.hadoop.ape.hcfs.CacheAwareFileSystem
+ * in hdfs configuration, for example, in hdfs-site.xml.
+ *
+ */
+public class CacheAwareFileSystem extends FileSystem {
+    private static final Logger LOG = LoggerFactory.getLogger(CacheAwareFileSystem.class);
+
+    private final static String SCHEME = "hdfs";
+
+    private URI uri;
+
+    /** The wrapped Hadoop File System. */
+    private org.apache.hadoop.fs.FileSystem hdfs;
+
+    private String locationPolicy;
+
+    @Override
+    public void initialize(URI name, Configuration conf) throws IOException {
+        uri = name;
+        hdfs = new DistributedFileSystem();
+        hdfs.initialize(name, conf);
+        LOG.info("fs initialized.");
+        
+        locationPolicy = getConf().get(
+                Constants.CONF_KEY_FS_APE_HCFS_BLOCK_LOCATION_POLICY,
+                Constants.CACHE_LOCATION_POLICY_MERGING_HDFS);
+        LOG.info("block location policy: {}", locationPolicy);
+    }
+
+    @Override
+    public BlockLocation[] getFileBlockLocations(FileStatus file, long start, long len) throws IOException {
+        if (file == null) {
+            throw new NullPointerException();
+        }
+
+        return getFileBlockLocations(file.getPath(), start, len);
+    }
+
+    @Override
+    public BlockLocation[] getFileBlockLocations(Path path, long start, long len) throws IOException {
+        if (path == null) {
+            throw new NullPointerException();
+        }
+        LOG.debug("getFileBlockLocations with: {}, start: {}, len: {}", path.toString(), start, len);
+
+        List<BlockLocation> result = new ArrayList<>();
+
+        if (start >= 0 && len > 0) {

```

```if (start >= 0 && len > 0)``` should we throw an exception if must matched case?
BTW, better to 
```
if (start < 0 || len < 0) {
  return result.toArray(new BlockLocation[0]);
}
```